### PR TITLE
Matching: document and refactor `mk_failaction_pos`

### PR DESCRIPTION
**Pull request by gasche**
`mk_failaction_pos` is one of the key operations in compilation of sum constructors. Its current code is fairly opaque -- it is hard to understand independently of the 2001 paper on pattern-matching compilation. I needed to understand the code in the context of fixing #7241, and it is reasonably likely that a future fix would touch this function -- so the fix would be difficult to review if the function is hard to understand.
The present PR documents this function and refactors it slightly for readability; it should not change its behavior in any way, and it should be reasonably easy to check this (this is just light, local code movement).

A possible review workflow would be as follows:

1. try to read the current definition of `mk_failaction_pos` in trunk and see what you can make sense of
2. read the definition in this PR (with all commits applied at once) and see if you find it easier to follow
3. once this is done, review each commit independently for correctness (it should clearly preserve the current behavior)

Checking the accuracy of the new documentation (instead of believing it and reviewing the phrasing only) requires more expertise in the pattern-matching compiler, I will try to get @maranget or @trefis to have a look. Other reviewers should not feel responsible for this.

### CODE REVIEW

File: lambda/matching.ml
Comment (lthls) [inline lambda/matching.ml]: Shouldn't this be `p_ctx` instead of `ctx` ?

Comment (gasche) [reply lambda/matching.ml]: It should, indeed, thanks! (One can check the previous behaviour of the `List.partition` call.) I rewrote this with an `as` pattern to be safer.

File: lambda/matching.ml
Comment (lthls) [inline lambda/matching.ml]: How would you feel about pattern-matching on the list instead of using an equality test ?
Or `List.is_empty`, since we have that now.

File: lambda/matching.ml
Comment (lthls) [inline lambda/matching.ml:2856]: `list_as_pat` is not a very informative name, and it is only used here. You could rename it to something like `or_pattern_from_list` and/or move the definition here.

**Comment by gasche** [comment]
I rebased the PR on trunk (this was non-trivial due to the debug printing changes that went in independently).

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml]: It is always better to go in the direction of the reviewer (so I will use `List.is_empty` here), but I don't find your proposals clearer than the current code. If you are thinking of the performance impact (neglectible here), maybe we should consider optimizing polymorphic comparisons to constants?

Comment (lthls) [reply lambda/matching.ml]: I'm not concerned about performance, but the previous code used pattern-matching, you have a pattern-matching that does the same thing 4 lines below, so I wondered why you chose to use the equality for this case only. (Some parts of the compiler re-export the `(=)` function with type `int -> int -> bool`, to prevent misuses of the polymorphic comparison, so I tend to be a bit defensive about it).

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml:2856]: I tried to please you and got pulled in a multi-file-change minor refactoring. I propose to send it as a separate PR to be evaluated on its own.

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml]: The consistency with the other line below is a good point. I rewrote the code to use a pattern-matching.

**Comment by gasche** [comment]
(I have gone through all of @lthls' comments. Thanks!)

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml:2856]: (The would-be PR is my branch at https://github.com/gasche/ocaml/commits/minor-patterns-refactoring . I am waiting for the present PR to be merged before submitting. )

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml:2847]: I believe this inverts the order of clauses from before, as the outer `fold_left` has been turned into a `concat_map`:

```ocaml
| A | B -> exit 1
| C | D -> exit 2

(* becomes *)

| C | D -> exit 2
| A | B -> exit 1
```

I don't believe this was an intentional or important property of the old code, and the new order should be semantically equivalent, so feel free to resolve this comment.

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]:
```suggestion
           fail patterns going to this exit. This is equivalent to
```

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]:
```suggestion
        (* We specialize the current context to the or-pattern of all
```

Nit: "Specialize at" or "specialize to" sounds more natural to me.

File: lambda/matching.ml
Comment (ncik-roberts) [reply lambda/matching.ml:2856]: +1

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml:2756]: In `mk_failaction_neg`, we use `partial` to identify that we can just return `None, Jumps.empty` and don't bother computing fail patterns. Do you know why we don't do that short-circuiting in `mk_failaction_pos` as well in the case that `fail_pats < match_context_rows`? If we trust `Total`, then it seems that short-circuiting would save us some needless work and keep contexts a bit smaller.

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]: Perhaps a missing detail is that the action for `fails` is always a raise to a larger exit number. (And that the "accuracy" mentioned below arises from skipping past unnecessary exit handlers.)

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]: The last sentence isn't quite accurate. Jumps from other handlers to that exit point might be `None`. It's more that jumps *from this switch* to that exit point must be `Some _`.

**Review by ncik-roberts — APPROVED** [review]
I read the old code and new code, and indeed I like the new code structure (and new comment) a lot more.

I've erred on the side on leaving nits about wording, as I'm a relative newcomer to this code and discussion may surface misunderstandings that I have.

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml:2756]: I don't know. It would also be possible to put this logic in the `if sig_complete then` check in the caller of `mk_failaction_pos` (`if sig_complete || partial = Total then ...`). (Something similar is done in `combine_variant`.) I would need to ask Luc if he considered this; and probably will. This could lead to a small code improvement in some cases of matching on an instance of a GADT type, where some cases can be eliminated by typing.

File: lambda/matching.ml
Comment (gasche) [reply lambda/matching.ml]: Attempted reformulation:

> For example, if a switch only accepts the [None] constructor, [mk_failaction_pos] generates a failure clause along with context information that the value reaching the failure clause must be [Some _].

**Comment by gasche** [comment]
@ncik-roberts thanks! I took your comments into account and rebased the PR.

(Hopefully this will encourage a maintainer, possibly @lthls, to approve and merge.)

**Review by lthls — APPROVED** [review]

**Comment by gasche** [comment]
Merged! Thank you both.

**Pull request merged**
